design-docs/Defer draft API.md

```diff
@@ -0,0 +1,852 @@
+reflected by merging them. The value of `extensions` is a map, and the same key can be present in several payloads. In
+that case it can be reflected by using a list as the value:
+
+```jsx
+// Payload 1:
+extensions: { "a": 1, "foo": "bar" }
+
+// Payload 2:
+extensions: { "foo": "baz" }
+```
+
+Would be received as:
+
+```kotlin
+// Emission 1:
+extensions = { a = 1, foo = bar }
+
+// Emission 2:
+extensions = { a = 1, foo = [bar, baz] }
+```
+
+**Impacts on ApolloInterceptor**
+
+Some implementations of `ApolloInterceptor` may assume that proceeding down the chain will return a single value,
+because that is always the case currently (except when using the normalized cache). Emitting several values in the Flow
+will break such implementations.
+
+### HttpEngine / HttpInterceptor / HttpResponse
+
+Here’s the current definitions of these 3 classes:
+
+```kotlin
+interface HttpEngine {
+  suspend fun execute(request: HttpRequest): HttpResponse
+  fun dispose()
+}
+
+interface HttpInterceptor {
+  suspend fun intercept(request: HttpRequest, chain: HttpInterceptorChain): HttpResponse
+  fun dispose() {}
+}
+
+class HttpResponse {
+  val statusCode: Int,
+  val headers: List<HttpHeader>,
+  val body: BufferedSource?,
+}
+```
+
+As we can see the contracts are designed around 1 full response being returned from a request, whereas when
+using `@defer` this is no longer appropriate: several chunks can be returned.
+
+To address this let’s augment `HttpResponse` with a Flow representing the chunks:
+
+```kotlin
+class HttpResponse {
+  val statusCode: Int,
+  val headers: List<HttpHeader>,
+  val body: BufferedSource?,
+  val chunks: Flow<HttpChunk>, // <- New!
+}
+
+class HttpChunk {
+  val headers: List<HttpHeader>,
+  val body: BufferedSource?,
+}
+```
+
+This introduces `HttpChunk` which is a light version of `HttpResponse`.
+
+Notes:
+
+- in the non-chunked (`@defer` not used) case, `chunks` will be an empty Flow
+- when `chunks` is not empty, the response `body` is `null`
+- any `HttpInterceptor` implementation that manipulates the response body won’t automatically work on chunks. Not sure
+  if this is a common use-case but it will break / not work as expected.
+
+### Http Cache
+
+The http cache can still be used in `@defer` scenarios.
+
+A response will be cached after all its chunks have been received.
+
+When a response comes from the cache, the chunks will be available immediately.
+
+### Batching
+
+At this point it is unclear whether batching in a `@defer` scenario is useful. For now, we can decide that batching is
+automatically opted-out for any operations using `@defer`.
+
+### Normalized Cache
+
+Several options are possible:
+
+**A. Incremental data is cached as it is being received**
+
+Pros:
+
+- When using `watch` the UI can be as reactive as possible and reflect the received data in real time
+
+Cons:
+
+- When using `query` with `FetchPolicy.CacheOnly` partial data can be returned, with no way to know this
```

Comment:
> The normalized cache knows of null vs absent so if data is not there yet, it's going to do a cache miss

Oh I didn't realize that, that's a good thing! So that's one less "con" for this option.

> Overall, I think the cache should act as a server that simply ignores all @defer directives.

I'm not exactly sure if you mean this in favor or option A or B?

Reply:
> I'm not exactly sure if you mean this in favor or option A or B?

I'm team A. Writers write data as soon as it's available. Readers read what's available from the cache. If someone tries to read a query containing `@defer` directives from the cache, the cache works as if `@defer` was not present and tries to resolve the deferred fragment as soon as possible. If it's not there, it's a cache miss

Reply:
👍  sounds fair! 

What about `watch` though? I feel like `watch` should emit the partial results as soon as they're received, because if not, it basically defeats the purpose of `@defer` when using the cache as the source of truth in a 'unidirectional data flow' architecture.

But that could mean some inconsistency: `watch` emits a partial result, while querying the cache directly doesn't return it.

Reply:
I think the question is not too much `A. store incremental data` vs `B. store only the final data`  but more how to handle reads. 

I think it makes sense to store the data as soon as it is received. Ideally, even store the individual payloads to not have to rewrite the initial ones all the time (requires the id but we will need it in all cases as discussed in the [comment below](https://github.com/apollographql/apollo-kotlin/pull/3858#pullrequestreview-877682891)).

> I feel like watch should emit the partial results as soon as they're received, because if not, it basically defeats the purpose of @defer when using the cache as the source of truth in a 'unidirectional data flow' architecture.

I'm not sure about this. Using `@defer` in a query that only hits the cache sounds like a red flag to me. The cache is fast, there should be no need to `@defer` fragments.

I _think_ `@defer` is another reason to move away from the current `watch()` (the one that supports fetchPolicy and refetchPolicy) and use `watch(data: Data?)` instead (the one that only reacts to store changes, without that initial fetch)

My favorite pattern for unidirectional data flows would be something like:

* each UI component has its own fragment 
* each UI component watches the store (CacheOnly, means we certainly need to introduce a watch() for fragments or else we'll have to wrap fragments in queries which is ok but cumbersome) 
* when the screen starts, a NetworkOnly query is made aggregating the various fragments, potentially using `@defer`

In that world, it's ok for the cache to ignore `@defer` completely (or maybe even throw). One issue with that model is that it makes it harder to communicate errors from the NetworkOnly query to the CacheOnly ones. But maybe once we can store more metadata in the cache we can store some error information there too?

The alternative is to start watching the cache while the initial fetch is still ongoing but I'm afraid this is trying to do "too many things at once"...

Reply:
That makes a lot of sense 🙏.

I'd say having UI components mapped to fragments matches well with the "spirit" of the spec - so having a way to observe fragments indeed would be useful.

So to summarize:
- data is stored in the cache as it's received incrementally
- querying the cache with fragments that are not in the cache (yet) leads to cache miss
- same is true with `watch(Data?)`: only 'full' data is emitted from the cache

I'll update the doc with this.